sdk/src/wallet/migration/mod.rs

```diff
@@ -0,0 +1,142 @@
+    // list.
+    &migrate_0::Migrate,
+];
+
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+pub struct MigrationVersion {
+    pub id: usize,
+    pub sdk_version: String,
+    pub date: time::Date,
+}
+
+impl std::fmt::Display for MigrationVersion {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{} - {} - {}", self.id, self.sdk_version, self.date)
+    }
+}
+
+#[async_trait]
+pub(crate) trait Migration {
+    const ID: usize;
+    const SDK_VERSION: &'static str;
+    const DATE: time::Date;
+
+    fn version() -> MigrationVersion {
+        MigrationVersion {
+            id: Self::ID,
+            sdk_version: Self::SDK_VERSION.to_string(),
+            date: Self::DATE,
+        }
+    }
+
+    #[cfg(feature = "storage")]
+    async fn migrate_storage(storage: &super::storage::Storage) -> Result<()>;
+
+    #[cfg(feature = "stronghold")]
+    async fn migrate_backup(storage: &crate::client::stronghold::StrongholdAdapter) -> Result<()>;
+}
+
+#[async_trait]
+trait DynMigration: Send + Sync {
+    fn version(&self) -> MigrationVersion;
+
+    #[cfg(feature = "storage")]
+    async fn migrate_storage(&self, storage: &super::storage::Storage) -> Result<()>;
+
+    #[cfg(feature = "stronghold")]
+    async fn migrate_backup(&self, storage: &crate::client::stronghold::StrongholdAdapter) -> Result<()>;
+}
+
+#[async_trait]
+impl<T: Migration + Send + Sync> DynMigration for T {
+    fn version(&self) -> MigrationVersion {
+        T::version()
+    }
+
+    #[cfg(feature = "storage")]
+    async fn migrate_storage(&self, storage: &super::storage::Storage) -> Result<()> {
+        let version = self.version();
+        log::info!("Migrating to version {}", version);
+        T::migrate_storage(storage).await?;
+        storage.set(MIGRATION_VERSION_KEY, version).await?;
+        Ok(())
+    }
+
+    #[cfg(feature = "stronghold")]
+    async fn migrate_backup(&self, storage: &crate::client::stronghold::StrongholdAdapter) -> Result<()> {
+        use crate::client::storage::StorageProvider;
+
+        let version = self.version();
+        log::info!("Migrating backups to version {}", version);
+        T::migrate_backup(storage).await?;
+        storage
+            .insert(
+                MIGRATION_VERSION_KEY.as_bytes(),
+                serde_json::to_string(&version)?.as_bytes(),
+            )
+            .await?;
+        Ok(())
+    }
+}
+
+#[cfg(feature = "storage")]
+pub async fn migrate_storage(storage: &super::storage::Storage) -> Result<()> {
+    let last_migration = storage.get::<MigrationVersion>(MIGRATION_VERSION_KEY).await?;
+    if last_migration
+        .as_ref()
+        .map(|m| m.id == LatestMigration::ID)
+        .unwrap_or_default()
+    {
+        return Ok(());
+    }
+    let next_migration = last_migration.map(|m| m.id + 1).unwrap_or_default();
```

Comment:
Do we care about the edge case where you'd be using a recent migration with an old SDK? Next step would panic?

Reply:
Hm. I suppose we could create an error for that.